packages/backend-storage/src/factory.ts

```diff
@@ -54,7 +66,63 @@ class AmplifyStorageFactory
         outputStorageStrategy
       );
     }
-    return constructContainer.getOrCompute(this.generator) as AmplifyStorage;
+    const amplifyStorage = constructContainer.getOrCompute(
+      this.generator
+    ) as AmplifyStorage;
+
+    this.generateAndAttachAccessPolicies(
+      getInstanceProps,
+      amplifyStorage.resources.bucket
+    );
+
+    return amplifyStorage;
+  };
+
+  private generateAndAttachAccessPolicies = (
+    getInstanceProps: ConstructFactoryGetInstanceProps,
+    bucket: IBucket
+  ) => {
+    const { ssmEnvironmentEntriesGenerator } = getInstanceProps;
+    const accessDefinition = this.props.access?.(storageAccessBuilder) || {};
+
+    const accessMap: Map<ResourceAccessAcceptor, Permission[]> = new Map();
+
+    Object.entries(accessDefinition).forEach(
+      ([s3Prefix, accessPermissions]) => {
+        accessPermissions.forEach((permission) => {
+          const resourceAccessAcceptor =
+            permission.getResourceAccessAcceptor(getInstanceProps);
+          if (!accessMap.has(resourceAccessAcceptor)) {
+            accessMap.set(resourceAccessAcceptor, []);
+          }
+          const prefix = `${s3Prefix}/${permission.resourceSuffix}`;
+          accessMap.get(resourceAccessAcceptor)?.push({
+            actions: permission.actions,
+            resources: [prefix],
+          });
+        });
+      }
+    );
+
+    const bucketPolicyFactory = new BucketPolicyFactory(bucket);
+
+    // TODO this is a bit unfortunate that this class has to maintain a cache of this value.
```

Comment:
This bit is interesting. I'm open to ideas

Reply:
I'd refactor this code to collaborating internal class. Doesn't have to be DI'ed, it can be `new`'d up in this class Ctor and maintain cache and references to necessary state.
(facing similar problem in client config PR, see https://github.com/aws-amplify/amplify-backend/blob/9278d3a00fac92bec7cb1809b09b89b2d1c90bb8/packages/backend/src/backend_factory.ts#L64-L66 )

Reply:
I suppose the other way to do this would be to inject an `SsmEnvironmentEntriesGeneratorFactory`, then this class can get an instance which can then maintain a cache of the value once computed. But maybe that's one too many levels of indirection?

Reply:
I'd wait with factory.

Reply:
Coming back to this, I think what I have here might be the best option for now. We can't initialize the internal class in the ctor because we need the bucket which is only created once `getInstance` is called. So we end up basically just repeating this logic to determine whether we need to initialize the internal class or not.

Reply:
Actually, I think I can move all of the policy generation logic to the construct generator and this problem goes away because the platform already guarantees that the generator is only called once